### PR TITLE
feat(warehouse): handle bucket config change

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

The wareHouse should be able to process the staging files even if the staging files are distributed to different Object Storage Configurations updated by the customer.
Spec: https://www.notion.so/rudderstacks/Bucket-config-changes-d61a41d111864f9a844a074b8595a8b2

## Notion Ticket
https://www.notion.so/rudderstacks/Support-bucket-config-changes-warehouse-d15703e536c348ab9d64a76f9fc37b11

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/query_builder.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		d.sourceID,
+		d.destinationID,
+		d.startStagingFileID,
+		d.endStagingFileID,
+	}...)
+	if err == sql.ErrNoRows {
+		err = nil
+		return
+	}
+	if err != nil && err != sql.ErrNoRows {
```
Comment (Jayachand) [inline warehouse/query_builder.go:41]: err != sql.ErrNoRows check not required

Comment (achettyiitr) [reply warehouse/query_builder.go]: Done.

File: warehouse/slave.go [hunk 2]
```diff
@@ -352,9 +352,19 @@ func processStagingFile(job PayloadT, workerIndex int) (loadFileUploadOutputs []
...
+	err = jobRun.downloadStagingFile(jobRun.job.DestinationConfig)
 	if err != nil {
-		return loadFileUploadOutputs, err
+		// If error occurs with the current config
+		// We retry with the revision config if it is present
+		if jobRun.job.DestinationRevisionConfig != nil {
+			pkgLogger.Infof("[WH]: Starting downloading staging file with revision config for StagingFileID: %d, StagingFileLocation: %s, whIdentifier: %s", job.StagingFileID, job.StagingFileLocation, jobRun.whIdentifier)
+			err = jobRun.downloadStagingFile(jobRun.job.DestinationRevisionConfig)
+			if err != nil {
+				return loadFileUploadOutputs, err
```
Comment (achettyiitr) [inline warehouse/slave.go:363]: Send some stats around this if both of the case will fail. Also log both the revisionId and current config.

Comment (achettyiitr) [reply warehouse/slave.go:363]: From @abhimanyubabbar

File: warehouse/utils/utils.go [hunk 3]
```diff
@@ -934,3 +937,27 @@ func GetLoadFilePrefix(timeWindow time.Time, warehouse WarehouseT) (timeWindowFo
 	}
 	return timeWindowFormat
 }
+
+func GetRequestWithTimeout(url string, timeout time.Duration) ([]byte, int, error) {
```
Comment (achettyiitr) [inline warehouse/utils/utils.go:941]: Added context. From @abhimanyubabbar

Comment (achettyiitr) [reply warehouse/utils/utils.go:941]: Use context.TODO() for now. But add a comment to change this later. Use application context going forward.

File: warehouse/slave.go (hunk 2 above)
Comment (achettyiitr) [reply warehouse/slave.go:363]: Done.
Added this stat worker_processing_download_staging_file_failed.

File: warehouse/utils/utils.go (hunk 3 above)
Comment (achettyiitr) [reply warehouse/utils/utils.go:941]: Done.

File: warehouse/slave.go [hunk 4]
```diff
@@ -352,9 +374,20 @@ func processStagingFile(job PayloadT, workerIndex int) (loadFileUploadOutputs []
 	jobRun.setStagingFileDownloadPath(workerIndex)
 
 	// This creates the file, so on successful creation remove it
-	err = jobRun.downloadStagingFile()
+	err = jobRun.downloadStagingFile(jobRun.job.CurrentDestinationConfig, job.CurrentUseRudderStorage)
```
Comment (abhimanyubabbar) [inline warehouse/slave.go:377]: Ideally a jobRun owns all the data necessary for performing operations around staging file. I was thinking that why we need to expose the logic to the caller by creating an opening for ability to supply the current destination config and use rudder storage options. Should we encapsulate it in the function itself ?

Comment (abhimanyubabbar) [reply warehouse/slave.go:377]: Wanted to hear your thoughts on it ?

File: warehouse/upload.go [hunk 5]
```diff
@@ -1622,6 +1624,50 @@ func (job *UploadJobT) getLoadFilesTableMap() (loadFilesMap map[tableNameT]bool,
...
 }
 
+func (job *UploadJobT) destinationRevisionIDMap() (revisionIDMap map[string]backendconfig.DestinationT, err error) {
+	revisionIDMap = make(map[string]backendconfig.DestinationT)
+	revisionIDs, err := distinctDestinationRevisionIdsFromStagingFiles(struct {
+		sourceID           string
+		destinationID      string
+		startStagingFileID int64
+		endStagingFileID   int64
+	}{sourceID: job.warehouse.Source.ID, destinationID: job.warehouse.Destination.ID, startStagingFileID: job.upload.StartStagingFileID, endStagingFileID: job.upload.EndStagingFileID})
```
Comment (abhimanyubabbar) [inline warehouse/upload.go:1634]: Can we decouple and create the variable beforehand for better readability ?

File: warehouse/upload.go [hunk 6]
```diff
@@ -1622,6 +1624,50 @@ func (job *UploadJobT) getLoadFilesTableMap() (loadFilesMap map[tableNameT]bool,
...
+	for _, revisionID := range revisionIDs {
+		// No need to make config backend api call for the current config
+		if revisionID == job.warehouse.Destination.RevisionID {
+			revisionIDMap[revisionID] = job.warehouse.Destination
+			continue
+		}
+
+		urlStr := fmt.Sprintf("%s/workspaces/destinationHistory/%s", configBackendURL, revisionID)
+
+		pkgLogger.Infof("[WH]: Get request Started for Dest revisionID %s", revisionID)
```
Comment (abhimanyubabbar) [inline warehouse/upload.go:1651]: Do we need these transient message logs for every revisionID ?

File: warehouse/utils/utils.go [hunk 7]
```diff
@@ -934,3 +948,27 @@ func GetLoadFilePrefix(timeWindow time.Time, warehouse WarehouseT) (timeWindowFo
 	}
 	return timeWindowFormat
 }
+
+func GetRequestWithTimeout(ctx context.Context, url string, timeout time.Duration) ([]byte, int, error) {
```
Comment (abhimanyubabbar) [inline warehouse/utils/utils.go:952]: Why we need to add responseCode to the function signature ? Wouldn't []byte and error be enough ? Wdyt ?

Comment (achettyiitr) [reply warehouse/utils/utils.go:952]: Done.

File: warehouse/upload.go (hunk 6 above)
Comment (achettyiitr) [reply warehouse/upload.go:1651]: Done.

File: warehouse/upload.go (hunk 5 above)
Comment (achettyiitr) [reply warehouse/upload.go:1634]: Done.

File: warehouse/query_builder.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
+package warehouse
+
+import (
+	"database/sql"
+	"fmt"
+
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+)
+
+func distinctDestinationRevisionIdsFromStagingFiles(d struct {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/query_builder.go:10]: Done.

File: warehouse/slave.go (hunk 4 above)
Comment (achettyiitr) [reply warehouse/slave.go]: Done.

File: warehouse/slave.go [hunk 9]
```diff
@@ -101,40 +119,58 @@ func (job *PayloadT) getFileManager() (filemanager.FileManager, error) {
...
-		panic(err)
-	}
-
 	job := jobRun.job
-	downloader, err := jobRun.job.getFileManager()
-	if err != nil {
-		pkgLogger.Errorf("[WH]: Failed to initialize downloader")
-		return err
-	}
+	downloadTask := func(config interface{}, useRudderStorage bool) (err error) {
```
Comment (abhimanyubabbar) [inline warehouse/slave.go:127]: Good job with using the inline function. Very concise and clear 🎉 🎈

File: warehouse/slave.go [hunk 10]
```diff
@@ -101,40 +119,58 @@ func (job *PayloadT) getFileManager() (filemanager.FileManager, error) {
...
 	}
-	file.Close()
-	timer.End()
 
-	fi, err := os.Stat(filePath)
+	err := downloadTask(job.CurrentDestinationConfig, job.CurrentUseRudderStorage)
 	if err != nil {
-		pkgLogger.Errorf("[WH]: Error getting file size of downloaded staging file: ", err)
-		return err
+		if job.StagingDestinationRevisionID != job.CurrentDestinationRevisionID && job.StagingDestinationConfig != nil {
```
Comment (abhimanyubabbar) [inline warehouse/slave.go:163]: I would wrap this up in a condition, to allow for better comprehension.





```suggestion
	if ( pickupStagingFileBucket(job) ) {
            err = downloadTask()
}
```

Comment (abhimanyubabbar) [reply warehouse/slave.go:163]: Wdyt ?

File: warehouse/slave.go [hunk 11]
```diff
@@ -354,7 +394,18 @@ func processStagingFile(job PayloadT, workerIndex int) (loadFileUploadOutputs []
...
 	if err != nil {
-		return loadFileUploadOutputs, err
+		// If error occurs with the current config and current revision is different from staging revision
+		// We retry with the staging revision config if it is present
+		if job.StagingDestinationRevisionID != job.CurrentDestinationRevisionID && job.StagingDestinationConfig != nil {
+			pkgLogger.Infof("[WH]: Starting processing staging file with revision config for StagingFileID: %d, CurrentDestinationRevisionID: %s, StagingDestinationRevisionID: %s, whIdentifier: %s", job.StagingFileID, job.CurrentDestinationRevisionID, job.StagingDestinationRevisionID, jobRun.whIdentifier)
+			err = jobRun.downloadStagingFile()
+			if err != nil {
+				job.sendDownloadStagingFileFailedStat()
+				return loadFileUploadOutputs, err
```
Comment (abhimanyubabbar) [inline warehouse/slave.go:404]: We don't need this anymore right ?

**Review by abhimanyubabbar — CHANGES_REQUESTED** [review]

File: warehouse/slave.go (hunk 11 above)
Comment (achettyiitr) [reply warehouse/slave.go:404]: Oh sorry. Completely missed removing this.

File: warehouse/slave.go (hunk 10 above)
Comment (achettyiitr) [reply warehouse/slave.go:163]: Makes sense. Will add test for that as well.

File: warehouse/slave.go (hunk 9 above)
Comment (achettyiitr) [reply warehouse/slave.go:127]: TYSM.

**Review by abhimanyubabbar — APPROVED** [review]
I think the changes look good 👍🏼 . There is good test coverage being added around this. Kudos 🎉

**Review by Jayachand — APPROVED** [review]
LGTM

**Pull request merged**